dockeradapter/filterdockerstream.go

```diff
@@ -40,46 +47,67 @@ const (
 // the prefix header bytes followed by the number of bytes specified in the prefix
 // size bytes.
 func FilterDockerStream(reader io.Reader, streamType int) ([]string, error) {
-	logging.Stderr("[filterdockerstream] filtering reader for steam type: ", streamType)
-	var (
-		result        = []string{}
-		prefix        = make([]byte, stdWriterPrefixLen)
-		payloadLength int
-		// TODO how come I have to declare this rather than use the assignment
-		// operator in the loop below? Doing so causes the function to return
-		// and empty slice?
-		bytesReturned int
-	)
-	_, err := reader.Read(prefix)
+	logging.Stderr("[FilterDockerstream] filtering reader for steam type: %d", streamType)
+	defer logging.Stderr("[FilterDockerstream] leaving")
+	result := []string{}
+	h, err := readFrameHeader(reader)
 	for err == nil {
-		payloadLength = int(binary.BigEndian.Uint32(prefix[stdWriterSizeIndex:]))
-		payload := make([]byte, payloadLength)
-		bytesReturned, err = reader.Read(payload)
+		payload, err := readFramePayload(reader, h)
 		if err != nil {
-			return nil, fmt.Errorf("Error reading payload")
-		}
-		if bytesReturned != payloadLength {
-			return nil,
-				fmt.Errorf("Bytes returned from stream [%d] does not match the length specified in the header [%d]",
-					bytesReturned,
-					payloadLength,
-				)
+			logging.Stderr("[FilterDockerstream] error reading payload: %s", err)
+			return result, err
 		}
-		lines := bytes.Split(payload, []byte("\n"))
-		// Only extract the requested stream type
-		if int(prefix[0]) == streamType {
+		// Discard payload if streamType doesn't match requested
+		if h.streamType == streamType {
+			lines := bytes.Split(payload, []byte("\n"))
 			for _, line := range lines {
 				if len(line) != 0 {
-					logging.Stderr("[filterdockerstream] read line from stream: ", string(line))
+					logging.Stderr("[FilterDockerstream] extracted line: %s", string(line))
 					result = append(result, string(line))
 				}
 			}
 		}
-		_, err = reader.Read(prefix)
-		if err != nil {
-			continue
+		h, err = readFrameHeader(reader)
+		if err == io.EOF {
+			break
 		}
 	}
-
 	return result, nil
 }
+
+// readFrameHeader returns a Docker stream frameHeader from the supplied Reader
+// See - https://docs.docker.com/engine/api/v1.29/#operation/ContainerAttach
+func readFrameHeader(reader io.Reader) (*frameHeader, error) {
+	logging.Stderr("[readFrameHeader] reading header")
+	prefix := make([]byte, stdWriterPrefixLen)
+
+	_, err := reader.Read(prefix)
+	if err != nil {
+		logging.Stderr("[readFrameHeader] error thrown reading frameHeader: %s", err)
```

Comment:
I'm leaving it to the caller to handle the EOF, adding some more info to the comment